File: webtool/server/lib/project.js
"use strict";

// Workspace project-state helper (new for this tool, not a port). Each
// project lives under webtool/workspace/<name>/ and owns:
//   unpack/                 - NitroPacker unpack output, never mutated after
//   translations/           - per-category split CSVs, the translation "save file"
//   build/                  - reinsert output (full copy of unpack/ with only
//                              translated .mes files overwritten)
//   output/                 - packed .nds output
//   project-state.json      - { romPath, projectName, createdAt, extractedAt }

const fs = require("fs");
const path = require("path");
const { execFile } = require("child_process");

const WORKSPACE_ROOT = path.resolve(__dirname, "..", "..", "workspace");
const REPO_ROOT = path.resolve(__dirname, "..", "..", "..");
const NITROPACKER_BIN = path.join(REPO_ROOT, "NitroPacker");

const NAME_RE = /^[A-Za-z0-9_-]+$/;

function assertValidName(name) {
  if (typeof name !== "string" || !NAME_RE.test(name)) {
    throw new Error(`invalid project name: ${JSON.stringify(name)} (use letters/digits/_/- only)`);
  }
}

function projectDir(name) {
  assertValidName(name);
  return path.join(WORKSPACE_ROOT, name);
}

function unpackDir(name) {
  return path.join(projectDir(name), "unpack");
}

function scriptDir(name) {
  return path.join(unpackDir(name), "data", "Script");
}

function csvDir(name) {
  return path.join(projectDir(name), "translations");
}

function originalRomPath(name) {
  return path.join(projectDir(name), "original.nds");
}

function buildDir(name) {
  return path.join(projectDir(name), "build");
}

function buildScriptDir(name) {
  return path.join(buildDir(name), "data", "Script");
}

function outputDir(name) {
  return path.join(projectDir(name), "output");
}

function statePath(name) {
  return path.join(projectDir(name), "project-state.json");
}

function readState(name) {
  const p = statePath(name);
  if (!fs.existsSync(p)) return null;
  return JSON.parse(fs.readFileSync(p, "utf-8"));
}

function writeState(name, patch) {
  fs.mkdirSync(projectDir(name), { recursive: true });
  const prev = readState(name) || {};
  const next = { ...prev, ...patch, projectName: name };
  fs.writeFileSync(statePath(name), JSON.stringify(next, null, 2));
  return next;
}

function listProjects() {
  if (!fs.existsSync(WORKSPACE_ROOT)) return [];
  return fs
    .readdirSync(WORKSPACE_ROOT, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .map((d) => d.name)
    .filter((name) => NAME_RE.test(name))
    .map((name) => readState(name))
    .filter(Boolean);
}

/** Find the NitroPacker project JSON file inside a project's unpack/build dir. */
function findProjectJson(dir) {
  if (!fs.existsSync(dir)) return null;
  const hit = fs.readdirSync(dir).find((f) => f.toLowerCase().endsWith(".json"));
  return hit ? path.join(dir, hit) : null;
}

/** Run the compiled NitroPacker CLI binary with the given argv, Promise-wrapped. */
function runNitroPacker(args) {
  return new Promise((resolve, reject) => {
    execFile(NITROPACKER_BIN, args, { maxBuffer: 64 * 1024 * 1024 }, (err, stdout, stderr) => {
      if (err) {
        reject(new Error(`NitroPacker ${args.join(" ")} failed: ${err.message}\n${stdout}\n${stderr}`));
        return;
      }
      resolve({ stdout, stderr });
    });
  });
}

module.exports = {
  WORKSPACE_ROOT,
  NITROPACKER_BIN,
  runNitroPacker,
  assertValidName,
  projectDir,
  unpackDir,
  scriptDir,
  csvDir,
  originalRomPath,
  buildDir,
  buildScriptDir,
  outputDir,
  statePath,
  readState,
  writeState,
  listProjects,
  findProjectJson,
};

File: webtool/server_py/project.py
"""
Workspace project-state helper (new for this tool, not a port of any
analysis/*.py script). Each project lives under webtool/workspace/<name>/
and owns:
  unpack/                 - NitroPacker unpack output, never mutated after
  translations/           - per-category split CSVs, the translation "save file"
  build/                  - reinsert output (full copy of unpack/ with only
                             translated .mes files overwritten)
  output/                 - packed .nds output
  project-state.json      - { romPath, projectName, createdAt, extractedAt }

Python port of webtool/server/lib/project.js - see that file's history for
the original design notes.
"""
import json
import os
import re
import subprocess

HERE = os.path.dirname(os.path.abspath(__file__))
WORKSPACE_ROOT = os.path.abspath(os.path.join(HERE, "..", "workspace"))
REPO_ROOT = os.path.abspath(os.path.join(HERE, "..", ".."))
NITROPACKER_BIN = (
    os.path.join(REPO_ROOT, "NitroPacker.exe")
    if os.path.exists(os.path.join(REPO_ROOT, "NitroPacker.exe"))
    else os.path.join(REPO_ROOT, "NitroPacker")
)

NAME_RE = re.compile(r"^[A-Za-z0-9_-]+$")


def assert_valid_name(name):
    if not isinstance(name, str) or not NAME_RE.match(name):
        raise ValueError(f"invalid project name: {name!r} (use letters/digits/_/- only)")


def project_dir(name):
    assert_valid_name(name)
    return os.path.join(WORKSPACE_ROOT, name)


def unpack_dir(name):
    return os.path.join(project_dir(name), "unpack")


def script_dir(name):
    return os.path.join(unpack_dir(name), "data", "Script")


def csv_dir(name):
    # Translation CSVs are NOT a per-project copy - they live in the repo
    # root's translations/, the same directory analysis/mes_translate_*.py's
    # CLI oracle reads/writes. A prior design kept a second copy under
    # workspace/<name>/translations/ manually synced to root, and that
    # "두 벌 관리" split was a repeated source of silent data-loss bugs (see
    # ANALYSIS_NOTES.md's many "두 트리 diff" entries). name is validated but
    # otherwise unused here.
    assert_valid_name(name)
    return os.path.join(REPO_ROOT, "translations")


def original_rom_path(name):
    return os.path.join(project_dir(name), "original.nds")


def build_dir(name):
    return os.path.join(project_dir(name), "build")


def build_script_dir(name):
    return os.path.join(build_dir(name), "data", "Script")


def output_dir(name):
    return os.path.join(project_dir(name), "output")


def state_path(name):
    return os.path.join(project_dir(name), "project-state.json")


def read_state(name):
    p = state_path(name)
    if not os.path.exists(p):
        return None
    with open(p, encoding="utf-8") as f:
        return json.load(f)


def write_state(name, patch):
    os.makedirs(project_dir(name), exist_ok=True)
    prev = read_state(name) or {}
    next_state = {**prev, **patch, "projectName": name}
    with open(state_path(name), "w", encoding="utf-8") as f:
        json.dump(next_state, f, indent=2, ensure_ascii=False)
    return next_state


def find_project_json(dir_path):
    """Find the NitroPacker project JSON file inside a project's unpack/build dir."""
    if not os.path.exists(dir_path):
        return None
    for f in os.listdir(dir_path):
        if f.lower().endswith(".json"):
            return os.path.join(dir_path, f)
    return None


def run_nitro_packer(args):
    """Run the compiled NitroPacker CLI binary with the given argv."""
    proc = subprocess.run(
        [NITROPACKER_BIN, *args],
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        raise RuntimeError(
            f"NitroPacker {' '.join(args)} failed: "
            f"exit {proc.returncode}\n{proc.stdout}\n{proc.stderr}"
        )
    return {"stdout": proc.stdout, "stderr": proc.stderr}
